File: lintcode/12.min-stack.py
class MinStack:
    def __init__(self):
        self.data_stack = []
        self.min_stack = []

    """
    @param: number: An integer
    @return: nothing
    """

    def push(self, number):
        if number <= min_stack[-1]:
            self.min_stack.append(number)
        self.data_stack.append(number)

    """
    @return: An integer
    """

    def pop(self):
        t = self.data_stack[-1]
        if t == self.min_stack[-1]:
            self.min_stack.pop()
        self.data_stack.pop()

    """
    @return: An integer
    """

    def min(self):
        return self.min_stack[-1]


if __name__ == '__main__':
    s = MinStack()
    s.push(1)
    s.pop()
    s.push(2)
    s.push(3)
    s.min()
    s.push(1)
    s.min()

File: lintcode/s/12.min-stack.cpp
class MinStack {
  stack<int> minStack;
  stack<int> dataStack;

public:
  MinStack() {
    // do intialization if necessary
  }

  /*
   * @param number: An integer
   * @return: nothing
   */
  void push(int number) {
    if (minStack.empty() || number <= minStack.top()) {
      minStack.push(number);
    }

    dataStack.push(number);
  }

  /*
   * @return: An integer
   */
  int pop() {
    if (dataStack.top() == minStack.top()) {
      minStack.pop();
    }
    int res = dataStack.top();

    dataStack.pop();
    return res;
  }

  /*
   * @return: An integer
   */
  int min() { return minStack.top(); }
};
